// @digdir/designsystemet-react/dist/esm/components/Pagination/index.js
'use client';
import { Pagination as Pagination$1 } from './Pagination.js';
import { PaginationContent } from './PaginationContent.js';
import PaginationEllipsis from './PaginationEllipsis.js';
import PaginationItem from './PaginationItem.js';
import { PaginationButton } from './PaginationButton.js';
import { PaginationPrevious, PaginationNext } from './PaginationNextPrev.js';
import PaginationRoot from './PaginationRoot.js';

const Pagination = Pagination$1;
Pagination.Root = PaginationRoot;
Pagination.Content = PaginationContent;
Pagination.Item = PaginationItem;
Pagination.Button = PaginationButton;
Pagination.Ellipsis = PaginationEllipsis;
Pagination.Previous = PaginationPrevious;
Pagination.Next = PaginationNext;
Pagination.Root.displayName = 'Pagination.Root';
Pagination.Content.displayName = 'Pagination.Content';
Pagination.Item.displayName = 'Pagination.Item';
Pagination.Button.displayName = 'Pagination.Button';
Pagination.Ellipsis.displayName = 'Pagination.Ellipsis';
Pagination.Previous.displayName = 'Pagination.Previous';
Pagination.Next.displayName = 'Pagination.Next';

export { Pagination, PaginationButton, PaginationContent, PaginationEllipsis, PaginationItem, PaginationNext, PaginationPrevious, PaginationRoot };
