// @digdir/designsystemet-react/dist/cjs/components/form/Radio/useRadio.js
'use client';
'use strict';

var React = require('react');
var useFormField = require('../useFormField.js');
var Group = require('./Group/Group.js');

/** Handles props for `Radio` in context with `Radio.Group` (and `Fieldset`) */
const useRadio = (props) => {
    const radioGroup = React.useContext(Group.RadioGroupContext);
    const { inputProps, readOnly, ...rest } = useFormField.useFormField(props, 'radio');
    return {
        ...rest,
        readOnly,
        inputProps: {
            ...inputProps,
            readOnly,
            type: 'radio',
            name: radioGroup?.name ?? props.name,
            required: radioGroup?.required,
            defaultChecked: radioGroup?.defaultValue
                ? radioGroup?.defaultValue === props.value
                : props.defaultChecked,
            checked: radioGroup?.value
                ? radioGroup?.value === props.value
                : props.checked,
            onClick: (e) => {
                if (readOnly) {
                    e.preventDefault();
                    return;
                }
                props?.onClick?.(e);
            },
            onChange: (e) => {
                if (readOnly) {
                    e.preventDefault();
                    return;
                }
                props?.onChange?.(e);
                radioGroup?.onChange?.(props.value);
            },
        },
    };
};

exports.useRadio = useRadio;
