// @digdir/designsystemet-react/dist/esm/components/form/Checkbox/useCheckbox.js
'use client';
import { useContext } from 'react';
import { useFormField } from '../useFormField.js';
import { CheckboxGroupContext } from './Group/Group.js';

/** Handles props for `Checkbox` in context with `Checkbox.Group` (and `Fieldset`) */
const useCheckbox = (props) => {
    const checkboxGroup = useContext(CheckboxGroupContext);
    const { inputProps, readOnly, ...rest } = useFormField(props, 'checkbox');
    return {
        ...rest,
        readOnly,
        inputProps: {
            ...inputProps,
            readOnly,
            type: 'checkbox',
            defaultChecked: checkboxGroup?.defaultValue
                ? checkboxGroup?.defaultValue.includes(props.value)
                : props.defaultChecked,
            checked: checkboxGroup?.value
                ? checkboxGroup?.value.includes(props.value)
                : props.checked,
            onClick: (e) => {
                if (readOnly) {
                    e.preventDefault();
                    return;
                }
                props?.onClick?.(e);
            },
            onChange: (e) => {
                if (readOnly) {
                    e.preventDefault();
                    return;
                }
                props?.onChange?.(e);
                checkboxGroup?.toggleValue(props.value);
            },
        },
    };
};

export { useCheckbox };
