// @digdir/designsystemet-react/dist/cjs/components/Spinner/Spinner.js
'use client';
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var lite = require('../../node_modules/clsx/dist/lite.js');
var getSize = require('../../utilities/getSize.js');
var useSynchronizedAnimation = require('../../hooks/useSynchronizedAnimation.js');

const sizeMap = {
    '2xs': 13,
    xs: 20,
    sm: 27,
    md: 40,
    lg: 56,
    xl: 79,
};
/**  Spinner component used for indicating busy or indeterminate loading */
const Spinner = ({ title, variant = 'default', className, style, ...rest }) => {
    const size = getSize.getSize(rest.size || 'md');
    const svgRef = useSynchronizedAnimation.useSynchronizedAnimation('fds-spinner-rotate-animation');
    const strokeRef = useSynchronizedAnimation.useSynchronizedAnimation('fds-spinner-stroke-animation');
    return (jsxRuntime.jsxs("svg", { className: lite.clsx('fds-spinner', `fds-spinner--${variant}`, className), style: { width: sizeMap[size], height: sizeMap[size], ...style }, viewBox: '0 0 50 50', ref: svgRef, ...rest, children: [jsxRuntime.jsx("title", { children: title }), jsxRuntime.jsx("circle", { className: lite.clsx('fds-spinner__background', variant === 'inverted' && 'fds-spinner__background--inverted'), cx: '25', cy: '25', r: '20', fill: 'none', strokeWidth: '5' }), jsxRuntime.jsx("circle", { className: lite.clsx(`fds-spinner__circle`), cx: '25', cy: '25', r: '20', fill: 'none', strokeWidth: '5', ref: strokeRef })] }));
};
Spinner.displayName = 'Spinner';

exports.Spinner = Spinner;
