// @digdir/designsystemet-react/dist/cjs/components/form/Checkbox/useCheckbox.js
'use client';
'use strict';

var React = require('react');
var useFormField = require('../useFormField.js');
var Group = require('./Group/Group.js');

/** Handles props for `Checkbox` in context with `Checkbox.Group` (and `Fieldset`) */
const useCheckbox = (props) => {
    const checkboxGroup = React.useContext(Group.CheckboxGroupContext);
    const { inputProps, readOnly, ...rest } = useFormField.useFormField(props, 'checkbox');
    return {
        ...rest,
        readOnly,
        inputProps: {
            ...inputProps,
            readOnly,
            type: 'checkbox',
            defaultChecked: checkboxGroup?.defaultValue
                ? checkboxGroup?.defaultValue.includes(props.value)
                : props.defaultChecked,
            checked: checkboxGroup?.value
                ? checkboxGroup?.value.includes(props.value)
                : props.checked,
            onClick: (e) => {
                if (readOnly) {
                    e.preventDefault();
                    return;
                }
                props?.onClick?.(e);
            },
            onChange: (e) => {
                if (readOnly) {
                    e.preventDefault();
                    return;
                }
                props?.onChange?.(e);
                checkboxGroup?.toggleValue(props.value);
            },
        },
    };
};

exports.useCheckbox = useCheckbox;
