// @digdir/designsystemet-react/dist/cjs/components/DropdownMenu/DropdownMenu.js
'use client';
'use strict';

var jsxRuntime = require('react/jsx-runtime');
var React = require('react');
var getSize = require('../../utilities/getSize.js');

const DropdownMenu = ({ open, onClose, placement = 'bottom-end', portal, children, ...rest }) => {
    const size = getSize.getSize(rest.size || 'md');
    const triggerRef = React.useRef(null);
    const [internalOpen, setInternalOpen] = React.useState(open ?? false);
    const anchorEl = triggerRef.current;
    const isControlled = typeof open === 'boolean';
    React.useEffect(() => {
        setInternalOpen(open ?? false);
    }, [open]);
    return (jsxRuntime.jsx(DropdownMenuContext.Provider, { value: {
            anchorEl,
            triggerRef,
            size,
            portal,
            placement,
            internalOpen,
            isControlled,
            onClose,
            setInternalOpen,
        }, children: children }));
};
const DropdownMenuContext = React.createContext({
    triggerRef: { current: null },
    size: 'md',
    anchorEl: null,
    internalOpen: false,
    setInternalOpen: () => { },
});
DropdownMenu.displayName = 'DropdownMenu';

exports.DropdownMenu = DropdownMenu;
exports.DropdownMenuContext = DropdownMenuContext;
