// @digdir/designsystemet-react/dist/cjs/components/Modal/index.js
'use client';
'use strict';

var ModalRoot = require('./ModalRoot.js');
var ModaContent = require('./ModaContent.js');
var ModalDialog = require('./ModalDialog.js');
var ModalFooter = require('./ModalFooter.js');
var ModalHeader = require('./ModalHeader.js');
var ModalTrigger = require('./ModalTrigger.js');

const Modal = ModalDialog.ModalDialog;
Modal.Root = ModalRoot.ModalRoot;
Modal.Content = ModaContent.ModalContent;
Modal.Footer = ModalFooter.ModalFooter;
Modal.Header = ModalHeader.ModalHeader;
Modal.Trigger = ModalTrigger.ModalTrigger;
Modal.Dialog = ModalDialog.ModalDialog;
Modal.Root.displayName = 'Modal.Root';
Modal.Dialog.displayName = 'Modal.Dialog';
Modal.Footer.displayName = 'Modal.Footer';
Modal.Header.displayName = 'Modal.Header';
Modal.Content.displayName = 'Modal.Content';
Modal.Trigger.displayName = 'Modal.Trigger';

exports.ModalContent = ModaContent.ModalContent;
exports.ModalDialog = ModalDialog.ModalDialog;
exports.ModalFooter = ModalFooter.ModalFooter;
exports.ModalHeader = ModalHeader.ModalHeader;
exports.ModalTrigger = ModalTrigger.ModalTrigger;
exports.Modal = Modal;
