// @digdir/designsystemet-react/dist/cjs/components/form/Search/useSearch.js
'use client';
'use strict';

var React = require('react');
var useFormField = require('../useFormField.js');
var FieldsetContext = require('../Fieldset/FieldsetContext.js');
var getSize = require('../../../utilities/getSize.js');

/** Handles props for `Search` in context with `Fieldset` */
const useSearch = (props) => {
    const fieldset = React.useContext(FieldsetContext.FieldsetContext);
    const { inputProps, readOnly, ...rest } = useFormField.useFormField(props, 'search');
    const size = fieldset?.size ?? getSize.getSize(props.size || 'md');
    return {
        ...rest,
        readOnly,
        size,
        inputProps: {
            ...inputProps,
            type: 'search',
            name: props.name ?? 'q',
            readOnly,
            onClick: (e) => {
                if (readOnly) {
                    e.preventDefault();
                    return;
                }
                props?.onClick?.(e);
            },
            onChange: (e) => {
                if (readOnly) {
                    e.preventDefault();
                    return;
                }
                props?.onChange?.(e);
            },
        },
    };
};

exports.useSearch = useSearch;
