// @digdir/designsystemet-react/dist/esm/components/Alert/Alert.js
'use client';
import { jsx, jsxs, Fragment } from 'react/jsx-runtime';
import { forwardRef } from 'react';
import { InformationSquareFillIcon, ExclamationmarkTriangleFillIcon, CheckmarkCircleFillIcon, XMarkOctagonFillIcon } from '@navikt/aksel-icons';
import { clsx } from '../../node_modules/clsx/dist/lite.js';
import { getSize } from '../../utilities/getSize.js';
import { Paragraph } from '../Typography/Paragraph/Paragraph.js';

const icons = {
    info: {
        Icon: InformationSquareFillIcon,
        title: 'Informasjon',
    },
    warning: { Icon: ExclamationmarkTriangleFillIcon, title: 'Advarsel' },
    success: { Icon: CheckmarkCircleFillIcon, title: 'Suksess' },
    danger: {
        Icon: XMarkOctagonFillIcon,
        title: 'Feil',
    },
};
const Alert = forwardRef(({ severity = 'info', elevated, iconTitle, children, className, ...rest }, ref) => {
    const size = getSize(rest.size || 'md');
    const { Icon, title } = icons[severity];
    return (jsx("div", { ref: ref, className: clsx('fds-alert', `fds-alert--${size}`, `fds-alert--${severity}`, elevated && `fds-alert--elevated`, className), ...rest, children: jsxs(Fragment, { children: [jsx(Icon, { title: iconTitle || title, className: 'fds-alert__icon' }), jsx(Paragraph, { asChild: true, size: size, className: 'fds-alert__content', children: jsx("span", { children: children }) })] }) }));
});
Alert.displayName = 'Alert';

export { Alert };
