// @digdir/designsystemet-react/dist/esm/components/form/Radio/useRadio.js
'use client';
import { useContext } from 'react';
import { useFormField } from '../useFormField.js';
import { RadioGroupContext } from './Group/Group.js';

/** Handles props for `Radio` in context with `Radio.Group` (and `Fieldset`) */
const useRadio = (props) => {
    const radioGroup = useContext(RadioGroupContext);
    const { inputProps, readOnly, ...rest } = useFormField(props, 'radio');
    return {
        ...rest,
        readOnly,
        inputProps: {
            ...inputProps,
            readOnly,
            type: 'radio',
            name: radioGroup?.name ?? props.name,
            required: radioGroup?.required,
            defaultChecked: radioGroup?.defaultValue
                ? radioGroup?.defaultValue === props.value
                : props.defaultChecked,
            checked: radioGroup?.value
                ? radioGroup?.value === props.value
                : props.checked,
            onClick: (e) => {
                if (readOnly) {
                    e.preventDefault();
                    return;
                }
                props?.onClick?.(e);
            },
            onChange: (e) => {
                if (readOnly) {
                    e.preventDefault();
                    return;
                }
                props?.onChange?.(e);
                radioGroup?.onChange?.(props.value);
            },
        },
    };
};

export { useRadio };
