// @digdir/designsystemet-react/dist/esm/components/form/Search/useSearch.js
'use client';
import { useContext } from 'react';
import { useFormField } from '../useFormField.js';
import { FieldsetContext } from '../Fieldset/FieldsetContext.js';
import { getSize } from '../../../utilities/getSize.js';

/** Handles props for `Search` in context with `Fieldset` */
const useSearch = (props) => {
    const fieldset = useContext(FieldsetContext);
    const { inputProps, readOnly, ...rest } = useFormField(props, 'search');
    const size = fieldset?.size ?? getSize(props.size || 'md');
    return {
        ...rest,
        readOnly,
        size,
        inputProps: {
            ...inputProps,
            type: 'search',
            name: props.name ?? 'q',
            readOnly,
            onClick: (e) => {
                if (readOnly) {
                    e.preventDefault();
                    return;
                }
                props?.onClick?.(e);
            },
            onChange: (e) => {
                if (readOnly) {
                    e.preventDefault();
                    return;
                }
                props?.onChange?.(e);
            },
        },
    };
};

export { useSearch };
