// @digdir/designsystemet-react/dist/esm/components/form/Switch/useSwitch.js
'use client';
import { useContext } from 'react';
import { useFormField } from '../useFormField.js';
import { CheckboxGroupContext } from '../Checkbox/Group/Group.js';

/** Handles props for `Switch` in context with `Checkbox.Group` (and `Fieldset`) */
const useSwitch = (props) => {
    const checkboxGroup = useContext(CheckboxGroupContext);
    const { inputProps, readOnly, ...rest } = useFormField(props, 'switch');
    const propsValue = props.value || '';
    return {
        ...rest,
        readOnly,
        inputProps: {
            ...inputProps,
            readOnly,
            type: 'checkbox',
            defaultChecked: checkboxGroup?.defaultValue
                ? checkboxGroup?.defaultValue.includes(propsValue)
                : props.defaultChecked,
            checked: checkboxGroup?.value
                ? checkboxGroup?.value.includes(propsValue)
                : props.checked,
            onClick: (e) => {
                if (readOnly) {
                    e.preventDefault();
                    return;
                }
                props?.onClick?.(e);
            },
            onChange: (e) => {
                if (readOnly) {
                    e.preventDefault();
                    return;
                }
                props?.onChange?.(e);
                checkboxGroup?.toggleValue(propsValue);
            },
        },
    };
};

export { useSwitch };
