// @digdir/designsystemet-react/dist/cjs/components/Pagination/index.js
'use client';
'use strict';

var Pagination$1 = require('./Pagination.js');
var PaginationContent = require('./PaginationContent.js');
var PaginationEllipsis = require('./PaginationEllipsis.js');
var PaginationItem = require('./PaginationItem.js');
var PaginationButton = require('./PaginationButton.js');
var PaginationNextPrev = require('./PaginationNextPrev.js');
var PaginationRoot = require('./PaginationRoot.js');

const Pagination = Pagination$1.Pagination;
Pagination.Root = PaginationRoot.default;
Pagination.Content = PaginationContent.PaginationContent;
Pagination.Item = PaginationItem.default;
Pagination.Button = PaginationButton.PaginationButton;
Pagination.Ellipsis = PaginationEllipsis.default;
Pagination.Previous = PaginationNextPrev.PaginationPrevious;
Pagination.Next = PaginationNextPrev.PaginationNext;
Pagination.Root.displayName = 'Pagination.Root';
Pagination.Content.displayName = 'Pagination.Content';
Pagination.Item.displayName = 'Pagination.Item';
Pagination.Button.displayName = 'Pagination.Button';
Pagination.Ellipsis.displayName = 'Pagination.Ellipsis';
Pagination.Previous.displayName = 'Pagination.Previous';
Pagination.Next.displayName = 'Pagination.Next';

exports.PaginationContent = PaginationContent.PaginationContent;
exports.PaginationEllipsis = PaginationEllipsis.default;
exports.PaginationItem = PaginationItem.default;
exports.PaginationButton = PaginationButton.PaginationButton;
exports.PaginationNext = PaginationNextPrev.PaginationNext;
exports.PaginationPrevious = PaginationNextPrev.PaginationPrevious;
exports.PaginationRoot = PaginationRoot.default;
exports.Pagination = Pagination;
